packages/vscode-extension/src/extension.ts

```diff
@@ -46,13 +51,48 @@ export function activate(context: ExtensionContext) {
     clientOptions,
   );
 
+  // Initialize singletons
+  LangugageClientManager.instance = new LangugageClientManager(client);
+  PersistentConnectionManager.instance = new PersistentConnectionManager();
+  ConnectionRepository.instance = new ConnectionRepository(
+    context.globalState,
+    context.secrets,
+  );
```

Comment:
is keeping a bunch of singletons around a common pattern you've seen in other vscode extensions? I worry that singletons generally make code harder to test since they add global state, have we considered other ways to achieve the same functionality?

Reply:
these have now largely been removed

I have a small class that returns an instance of globalState and secretManager.. these 2 objects are only available on the ExtensionContext object, which (as far as I can see) you can't access elsewhere during runtime outside of the activate function.. this getAppContext function I wrote is looking at me a bit funny.. but this does seem to be the proscribed method for data storage https://code.visualstudio.com/api/extension-capabilities/common-capabilities#data-storage